### ISSUE TITLE
Black output on some input/output buffer sizes

### ISSUE THREAD

**Issue by Gobra** [halide/Halide#3111]
I have quite simple filter, a nearest-neighbor scaler that looks like that:
```
Buffer<> input;
Func output;
Var x, y, c;

Param<float> sx = ...;
Param<float> sy= ...;
auto mx = cast<int>(x / sx);
auto my = cast<int>(y / sy);
output(x, y, c) = input(mx, my, c);
```

The input image I test the code with is `3464x1948` pixels. I have an UI with a slider that changes scale inputs and changes output buffer size according to the expected scaled size. The filter works fine until output buffer hits around `5500x3150` pixels - after around that number the output is simply black image. Than once I scale up to `11200x6300` pixels it works again.

I'm running it on Win10, OpenCL target, I have Halide compiled from master checked out 2018_06_21. Let me know if I can get you guys more info regarding the issue.

**Comment by steven-johnson** [comment]
Does it happen for CPU-only, or does it only misbehave for OpenCL?

**Comment by Gobra** [comment]
I've tested with OpenCL only.

The problem is I can not get it working with `target_.set_feature(Halide::Target::CPlusPlusMangling);` as I get following error:
```
File: d:\source\apps\shared\source\llvm\llvm-6.0\...\error.cpp
Line: 116
Expression: Err && "report_fatal_error called with success value"
```

Is there other way to force it to CPU? Actually, is there any debug option to have exact OpenCL device chosen by Halide to be printed to console/file or maybe even make it chose specific device?

**Comment by steven-johnson** [comment]
At this point it would probably be most helpful if you could post a complete working example that replicates the problem.

**Comment by Gobra** [comment]
I've rather big app, so I tried to reduce the code to some minimum and - surprise - it worked as expected. So I simplified my "big" code as much as possible why maintaining the issue and got the following: now I have a simple 1-function code like mentioned above that **works** and a bit more complicated one yet implementing the same formula that **does not** and produces black output on some sizes.

Now I've printed output functions for both with `.compile_to_lowered_stmt()` and got the same output except of function and parameters indices. I.e the formula is the same, but my "complicted" code has various places where params or functions might be copied and it seems that results the difference.

Since my "buggy" code is rather complex thing with C++ and JavaScript, it's rather pointless to post it all, @steven-johnson can think of something a might be from Halide that would help to solve this mystery? Some debug trace or maybe I can check something with debug version digging into Halide's internals?

**Comment by Gobra** [comment]
Just to clarify, I've attached two files. `scaler-direct` is printed function assembled in a few lines of code, `scaler-pipeline` is one build inside my code through JS/C++ interop and all the stuff I have in the app.

The files are actually HTMLs, Github doesn't support those, so I've renamed them to TXT. Rename back and compare, you'll see it's the same thing, yet `scaler-pipeline` produces black image on the dimensions I've mentioned in the first post.

[scaler-direct.txt](https://github.com/halide/Halide/files/2177840/scaler-direct.txt)
[scaler-pipeline.txt](https://github.com/halide/Halide/files/2177841/scaler-pipeline.txt)
